src/boot/efi/pe.c

```diff
@@ -224,6 +253,57 @@ static void pe_locate_sections(
                 }
 }
 
+static bool pe_use_this_dtb(
+                const PeSectionHeader section_table[],
+                size_t n_section_table,
+                void *base,
+                size_t dtb_offset,
+                size_t dtb_size,
+                size_t section_nb) {
+        EFI_STATUS err;
+
+        static const void *cached_base = NULL;
+        static const Device *cached_device = NULL;
+
+        const uint8_t *dtb = (const uint8_t *) base + dtb_offset;
+
+        err = devicetree_match(dtb, dtb_size);
+
+        /* Firmware does not provide the devicetree, so try matching against a list from .hwids section */
+        if (err == EFI_UNSUPPORTED) {
+                static const char *const hwids_section_name[] = { ".hwids", NULL };
+                PeSectionVector hwids_section = {};
+                pe_locate_sections(
+                                section_table,
+                                n_section_table,
+                                hwids_section_name,
+                                PTR_TO_SIZE(base),
+                                &hwids_section);
+                if (hwids_section.memory_size == 0)
+                        log_error_status(err, "HWIDs section is missing, no DT blob will be selected");
+                else {
+                        if (cached_base != base) {
+                                cached_base = base;
+                                cached_device = NULL;
+                                const uint8_t *hwids = (const uint8_t *) base + hwids_section.memory_offset;
+                                err = chid_match(hwids, hwids_section.memory_size, &cached_device);
+                        } else
+                                err = cached_device ? EFI_SUCCESS : EFI_NOT_FOUND;
+
+                        const char *compatible = device_get_compatible(cached_base, cached_device);
+                        if (err == EFI_SUCCESS && compatible)
+                                err = devicetree_match_by_compatible(dtb, dtb_size, compatible);
+                }
+        }
+
+        if (err == EFI_SUCCESS)
+                return true;
+        else if (err == EFI_INVALID_PARAMETER)
+                log_error_status(err, "Found bad DT blob in PE section %zu", section_nb);
```

Comment:
please move this up, i.e. right after the devicetree_match() call.

i.e. please do error handling error, and prefer early-exit over deep nesting.